hh-plus-plus.js

```diff
@@ -4426,6 +4455,191 @@ function simuFight(player, opponent) {
+        return {
+            scoreStr: "100%",
+            scoreClass: "plus"
+        };
+    } else if (Math.floor(1+player.hp/opponent.dmg)*2*player.dmg < opponent.hp) {
+        // guaranteed loss
+        return {
+            scoreStr: "0%",
+            scoreClass: "minus"
+        };
+    } else if (Math.ceil(opponent.hp/player.dmg-1)*2*opponent.dmg < player.hp) {
+        // guaranteed win
+        return {
+            scoreStr: "100%",
+            scoreClass: "plus"
+        };
+    }
+
+    // Amount of non-crit hits we can take without losing
+    const tolerableHits = Math.ceil(player.hp/opponent.dmg)-1;
+
+    let winChance = 0;
+    let loseChance = 0;
+    let playerCrits = 0;
+    let playerNormalHits = Math.ceil(opponent.hp/player.dmg);
+
+    if(logging) console.log('Probability calculation log for: ' + opponent.name);
+    do {
+        if(logging) console.log(' Scenario: ' + playerCrits + ' crits and ' + playerNormalHits + ' hits');
+        let scenarioLikelihood = calculateChance(playerCrits, playerNormalHits, player.critchance);
+        let overkillChance = calculateOverkillChance(playerCrits, playerNormalHits, player.critchance);
+        if(logging) console.log('  Scenario likelihood: ' + 100*scenarioLikelihood + ' % + ' + 100*overkillChance + ' % chance for overkill');
+        scenarioLikelihood += overkillChance;
+
+        let rounds = playerCrits + playerNormalHits;
+        let tolerableCrits = tolerableHits-rounds+1;
+        if(logging) console.log('  Opponent is allowed to crit ' + tolerableCrits + ' times');
+
+        if (tolerableCrits < 0) {
+            if(logging) console.log('  => impossible, we lose');
+            loseChance += scenarioLikelihood;
+        } else if (tolerableCrits >= rounds-1) {
+            if(logging) console.log ('  => guaranteed, we win');
+            winChance += scenarioLikelihood;
+        } else {
+            let opponentLikelihood = 0;
+            for(let i=0; i<=tolerableCrits; i++) {
+                let tmp = calculateChance(i, rounds-i-1, opponent.critchance);
+                if(logging) console.log('   probability for ' + i + ' crits and ' + (rounds-i-1) + ' hits: ' + 100*tmp + ' %');
+                opponentLikelihood += tmp;
+            }
+            if(logging) console.log('  ' + 100*opponentLikelihood + ' % chance that this condition is fulfilled');
+            if(logging) console.log('  => ' + 100*opponentLikelihood*scenarioLikelihood + ' % to win through this scenario');
+            winChance += opponentLikelihood*scenarioLikelihood;
+            loseChance += (1-opponentLikelihood)*scenarioLikelihood;
+        }
+
+        playerCrits++;
+        playerNormalHits-=2;
+    } while (playerNormalHits >= 0);
+
+    if(logging) console.log(100*winChance+ ' % chance to win vs. ' + 100*loseChance + ' % chance to lose => ' + 100*(winChance+loseChance) + ' % total coverage.');
+
+    return {
+        scoreStr: (100*winChance).toFixed(2) + '%',
```

Comment:
This causes display inconsistency with '100%' for guaranteed win shortcut and '100.00%' for the calculated probability. Prefer multiplying by 10,000, rounding and then dividing by 100?

Reply:
Note: the nRounding utility function does this for you.

Reply:
I don't think that nRounding would help there. It also uses toFixed in the end which will always add the digits. In fact there was the issue if you had less than 1k money that the script would add ".000" (should be fixed with 8dfb5d2).
To get rid of zeros at the end, you'd have to use the floating point directly (either with rounding or converting the string back) and hope there are no weird precision issues. Or you could do a string replacement to clean up the end, e.g. `.toFixed(2).replace(/\.?0*$/,'')`
The most straight forward solution would be to just return "100.00%"/"0.00%" on the shortcut for consistency.

However the "100.00%" aren't identical to an actual 100%. Even if it's extremely unlikely, there is a tiny chance to lose if you never crit and the opponent does a lot. The log should point that out:
![hhpp-100](https://user-images.githubusercontent.com/88989130/130365270-4dbe8c9a-1c08-41ec-bc25-ff980e2a1055.png)
One idea I had was to round towards 50% so calculation results would be in the range of 0.01% and 99.99% and a 0% or 100% would be an actual guaranteed result. nRounding would be useful for that to control the rounding. 
What do you think? This might be a bit too nitpicky. Most people probably won't care about that once in a billion chance to lose despite seeing a 100% prediction.

Reply:
Fair enough. For the prediction then, I would say just round down instead of towards 50%. Keep the expectations low.